packages/hop-node/src/1inch/1inch.ts

```diff
@@ -0,0 +1,347 @@
+  }
+
+  async getQuote (params: QuoteParams) {
+    const { fromTokenAddress, toTokenAddress, amount } = params
+    if (!this.chainId) {
+      throw new Error('chainId is required')
+    }
+    if (!fromTokenAddress) {
+      throw new Error('fromTokenAddrss is required')
+    }
+    if (!toTokenAddress) {
+      throw new Error('toTokenAddress is required')
+    }
+    if (!amount) {
+      throw new Error('amount is required')
+    }
+
+    const url = this.constructUrl('/quote', {
+      fromTokenAddress,
+      toTokenAddress,
+      amount
+    })
+    const result = await this.getJson(url)
+    if (!result.toTokenAmount) {
+      logger.error(result)
+      throw new Error('expected tx data')
+    }
+
+    const { toTokenAmount } = result
+
+    return toTokenAmount
+  }
+
+  async getAllowance (params: AllowanceParams) {
+    const { tokenAddress, walletAddress } = params
+    if (!this.chainId) {
+      throw new Error('chainId is required')
+    }
+    if (!tokenAddress) {
+      throw new Error('tokenAddress required')
+    }
+    if (!walletAddress) {
+      throw new Error('walletAddress is required')
+    }
+
+    const url = this.constructUrl('/approve/allowance', {
+      tokenAddress,
+      walletAddress
+    })
+    const result = await this.getJson(url)
+
+    if (result.allowance === undefined) {
+      logger.error(result)
+      throw new Error('expected tx data')
+    }
+
+    return result.allowance
+  }
+
+  async getApproveTx (params: ApproveParams) {
+    const { tokenAddress, amount } = params
+    if (!this.chainId) {
+      throw new Error('chainId is required')
+    }
+    if (!tokenAddress) {
+      throw new Error('tokenAddress required')
+    }
+    if (!amount) {
+      throw new Error('amount is required')
+    }
+
+    const url = this.constructUrl('/approve/transaction', {
+      amount,
+      tokenAddress
+    })
+    const result = await this.getJson(url)
+
+    if (!result.data) {
+      logger.error(result)
+      throw new Error('expected tx data')
+    }
+
+    const { data, to, value } = result
+
+    return {
+      data,
+      to,
+      value
+    }
+  }
+
+  async getSwapTx (params: SwapParams) {
+    const { fromTokenAddress, toTokenAddress, fromAddress, amount, slippage, destReceiver } = params
+    if (!this.chainId) {
+      throw new Error('chainId is required')
+    }
+    if (!fromTokenAddress) {
+      throw new Error('fromTokenAddrss is required')
+    }
+    if (!toTokenAddress) {
+      throw new Error('toTokenAddress is required')
+    }
+    if (!fromAddress) {
+      throw new Error('fromAddress is required')
+    }
+    if (!amount) {
+      throw new Error('amount is required')
+    }
+    if (!slippage) {
+      throw new Error('slippage is required')
+    }
+
+    const url = this.constructUrl('/swap', {
+      fromTokenAddress,
+      toTokenAddress,
+      amount,
+      fromAddress,
+      slippage,
+      destReceiver
+    })
+    const result = await this.getJson(url)
```

Comment:
What is the result when the returned amount is out of the slippage range defined? I think we should explicitly handle that here.

Reply:
The api returns this `cannot estimate` error if slippage param is too low
![DeepinScreenshot_select-area_20220301094949](https://user-images.githubusercontent.com/168240/156221719-c489279b-3be7-4a6d-b3f0-a78f444deaab.png)

but i'll add additional check as well

Reply:
the calculated slippage check has been [added](https://github.com/hop-protocol/hop/commit/1e3f2a74c4b1b772710515bd935e045ffd718a9d) and could use another set of eyes for sanity check